contracts/Vesting.sol

```diff
@@ -1,6 +1,131 @@
+        require(
+            investors_.length == amount_.length &&
+                investors_.length == allocation_.length,
+            "Error : Different arrays length"
+        );
+        uint256 sumOfAmount;
+        for (uint256 i; i < investors_.length; i++) {
+            require(
+                investors_[i] > address(0) && amount_[i] > 0,
+                "Error : 'investors_' or 'amount_' , equal to 0"
+            );
+            if (allocation_[i] == Allocation.Seed) {
+                allocationList[allocation_[i]][
+                    investors_[i]
+                ] = listOfBeneficiaries[investors_[i]];
+                listOfBeneficiaries[investors_[i]].intialReward += ((amount_[
+                    i
+                ] / 10000) * 1000);
+                listOfBeneficiaries[investors_[i]].balanceBase += amount_[i];
+            }
+            if (allocation_[i] == Allocation.Private) {
+                allocationList[allocation_[i]][
+                    investors_[i]
+                ] = listOfBeneficiaries[investors_[i]];
+                listOfBeneficiaries[investors_[i]].intialReward += ((amount_[
+                    i
+                ] / 10000) * 1500);
+                listOfBeneficiaries[investors_[i]].balanceBase += amount_[i];
+            }
+            sumOfAmount += amount_[i];
+        }
+        _token.safeTransferFrom(msg.sender, address(this), sumOfAmount);
+        emit AddInvestors(investors_, amount_);
+    }
+
+    function withdrawTokens() external override {
+        require(vestingStartDate > 0, "Error : first set 'vestingStartDate'");
+        require(
+            vestingCliff < block.timestamp,
+            "Error : wait until cliff period is end"
+        );
+        require(
+            listOfBeneficiaries[msg.sender].balanceBase > 0,
+            "Error : not enougth tokens"
+        );
+        require(
+            listOfBeneficiaries[msg.sender].rewardPaid <
+                listOfBeneficiaries[msg.sender].balanceBase,
+            "Error : not enougth tokens"
+        );
+        uint256 amount = listOfBeneficiaries[msg.sender].intialReward +
+            (_calculateUnlock(100));
+        require(amount > 0, "Error : 'amount' equal to 0");
+        listOfBeneficiaries[msg.sender].intialReward = 0;
+        listOfBeneficiaries[msg.sender].rewardPaid += amount;
+        _token.safeTransfer(msg.sender, amount);
+        emit Withdraw(msg.sender, amount);
+    }
+
+    function _calculateUnlock(uint256 percent) internal view returns (uint256) {
+        uint256 onePercentInTokens = (listOfBeneficiaries[msg.sender]
+            .balanceBase / 10000) * percent;
+        if (block.timestamp < vestingDuration) {
+            uint256 actualTime = (block.timestamp - vestingCliff) / 6 minutes;
+            return
+                (actualTime * onePercentInTokens) -
+                listOfBeneficiaries[msg.sender].rewardPaid;
+        } else if (block.timestamp > vestingDuration) {
```

Comment:
* you never check it's in tests

Reply:
i try calculate:
example
user A set next:

initialReward = 100
balanceBase = 1000

then user wait 50% vesting period and call withdraw tokens and have next:

onePercentInTokens = 10

50 * 10 = 500

(50 * 10) + 100 - 0 = user can claim 600 tokens when vesting is progress on 50% but in correct flow user must can claim only 550 tokens

100 initial + ( 50 * 10) ; 9 its one percent = 1000 - initalReward = 900/100 = 9


Next case user wait when claim is progress on 99%
and claim we have next

99 *  10 + 100 = 1090 tokens which is overflow which is more than the amount he should have received